### PR TITLE
repo-updater: Syncing improvements for Bitbucket Server

### PR DESCRIPTION
**Pull request by tsenart**
This PR brings the repository syncing improvements of #2025 to Bitbucket Server external services.

### CODE REVIEW

File: cmd/repo-updater/repos/bitbucketserver.go [hunk 1]
```diff
@@ -318,61 +325,136 @@ func newBitbucketServerConnection(config *schema.BitbucketServerConnection, cf h
...
-	perPage := 100
-	ch := make(chan *bitbucketserver.Repo, perPage)
+func (c *bitbucketServerConnection) excludes(r *bitbucketserver.Repo) bool {
+	name := r.Slug
+	if r.Project != nil {
+		name = r.Project.Key + "/" + name
+	}
+	return c.exclude[strings.ToLower(name)] ||
+		c.exclude[strconv.Itoa(r.ID)] ||
+		c.config.ExcludePersonalRepositories && r.IsPersonalRepository()
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/bitbucketserver.go]: I'm guessing the evaluation rules work out here, but this makes me uneasy. Can we make this last line its own subexpression to be explicit.

```suggestion
		(c.config.ExcludePersonalRepositories && r.IsPersonalRepository())
```

File: cmd/repo-updater/repos/bitbucketserver.go [hunk 2]
```diff
@@ -318,61 +325,136 @@ func newBitbucketServerConnection(config *schema.BitbucketServerConnection, cf h
...
 
-	return ch
+	seen := make(map[int]bool)
+	errs := new(multierror.Error)
+	var repos []*bitbucketserver.Repo
+
+	for r := range ch {
+		if r.err != nil {
+			errs = multierror.Append(errs, r.err)
+			continue
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/bitbucketserver.go]: `r.repos` can be non-empty even if `r.err != nil`, so you don't wanna continue here.

File: cmd/repo-updater/repos/bitbucketserver.go [hunk 3]
```diff
@@ -318,61 +325,136 @@ func newBitbucketServerConnection(config *schema.BitbucketServerConnection, cf h
...
+		}
+
+		for _, repo := range r.repos {
+			if !seen[repo.ID] && !c.excludes(repo) && repo.State == "AVAILABLE" {
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/bitbucketserver.go]: should we just put `repo.State == "AVAILABLE"` into the excludes function?

File: cmd/repo-updater/repos/gitlab.go [hunk 4]
```diff
@@ -344,7 +344,7 @@ func (c *gitlabConnection) listAllProjects(ctx context.Context) ([]*gitlab.Proje
 					// TODO(tsenart): When implementing dry-run, reconsider alternatives to return
 					// 404 errors on external service config validation.
 					if gitlab.IsNotFound(err) {
-						log15.Warn("skipping missing gitlab.projects entry:", "name", p.Name, "id", p.Id, "err", err)
+						log15.Warn("skipping missing gitlab.projects entry:", "name", p.Name, "id", fmt.Sprint(p.Id), "err", err)
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/gitlab.go]: why the Sprint here?

Comment (tsenart) [reply cmd/repo-updater/repos/gitlab.go]: As spoken offline, no need.

File: cmd/repo-updater/repos/bitbucketserver.go (hunk 3 above)
Comment (tsenart) [reply cmd/repo-updater/repos/bitbucketserver.go]: Let's do it.

File: cmd/repo-updater/repos/bitbucketserver.go (hunk 2 above)
Comment (tsenart) [inline cmd/repo-updater/repos/bitbucketserver.go]:
```suggestion
```

Comment (tsenart) [reply cmd/repo-updater/repos/bitbucketserver.go]: Resolved.

File: cmd/repo-updater/repos/bitbucketserver.go [hunk 5]
```diff
@@ -146,7 +149,7 @@ func bitbucketServerRepoInfo(config *schema.BitbucketServerConnection, repo *bit
 	return &protocol.RepoInfo{
 		Name: repoName,
 		ExternalRepo: &api.ExternalRepoSpec{
-			ID:          project + "/" + repo.Slug,
+			ID:          strconv.Itoa(repo.ID),
```
[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/repo-updater/repos/bitbucketserver.go]: Yes.

**Comment by tsenart** [comment]
> How was implementing this compared to GitLab (especially time to complete).

Both GitLab and Bitbucket Server took two days of work until PR was out.

> service type is a constant: bitbucketserver.ServiceType

Can we follow up on a PR that does this for all code hosts at once?

> service type mixes case bitbucketServer. We may have some assumptions somewhere that everything in the repos table is uppercased.

I think this is covered appropriately with tests already.

**Comment by tsenart** [comment]
@sqs: Please review CHANGELOG.

**Comment by tsenart** [comment]
@keegancsmith: I'm out for the day. Once @sqs approves, please merge the PR.

**Comment by tsenart** [comment]
@sqs: If you have some feedback on the CHANGELOG, please follow up with a PR to implement those suggestions.

**Comment by tsenart** [comment]
Damn, forgot to change the base branch, so this wasn't merged in master.

**Pull request merged**